### DISCUSSION TITLE
Variable unpacking for type concatenation with variadic tuple

### DISCUSSION THREAD

**Discussion by sylee957** [microsoft/pyright#4889] [Q&A] [answered]
If I have a function like

```python
from typing import reveal_type


def f() -> tuple[int, str, *tuple[bool, ...]]:
    return 1, "a", True, False
```

If I unpack as correct shape as `first, second, *rest`, the result is satisfactory
```python3
first, second, *rest = f()
reveal_type(first)  # Revealed type is "int"
reveal_type(second)  # Revealed type is "str"
reveal_type(rest)  # Revealed type is "list[bool]"
```

However when I unpack as `first, *rest`, I notice that the inference of `rest` is wrong as `list[str]`.
```python3
first, *rest = f()
reveal_type(first)  # Revealed type is "int"
reveal_type(rest)  # Revealed type is "list[str]"
```

However, if the function is
```python3
def f() -> tuple[int, str, *tuple[bool, bool]]:
    return 1, "a", True, False
```
where the unpack arguments are finitary, it correctly infers as
```python3
first, *rest = f()
reveal_type(first)  # Revealed type is "int"
reveal_type(rest)  # Revealed type is "list[str | bool]"
```
I'm using pyright 1.1.301

**Comment by erictraut** [comment, marked as answer]
> I notice that the inference of rest is wrong

You're creating an ambiguous situation for the type checker. The inferred type of `*rest` should be `list[str]` if `*tuple[bool, ...]` contains zero entries. If it contains more than one, the inferred type of `*rest` should be `list[str | bool]`. When you use `...` in a tuple, a type checker needs to make certain assumptions when unpacking. It gets even more complicated if the unpacked target is in the middle (e.g. `a, *b, c` = f()`).

That said, I agree that `list[str | bool]` is a better inference result in this case. I've updated the logic to produce this result.

When you remove the `...` and switch to `tuple[int, str, *tuple[bool, bool]]`, you remove the ambiguity. This is the equivalent of `tuple[int, str, bool, bool]`.
